sharding/shardset.go

```diff
@@ -56,20 +60,32 @@ func NewEmptyShardSet(fn HashFn) ShardSet {
 
 func newValidatedShardSet(shards []shard.Shard, fn HashFn) ShardSet {
 	ids := make([]uint32, len(shards))
-	for i := range ids {
-		ids[i] = shards[i].ID()
+	shardMap := make(map[uint32]shard.Shard, len(shards))
+	for i, shard := range shards {
+		ids[i] = shard.ID()
+		shardMap[shard.ID()] = shard
 	}
 	return &shardSet{
-		shards: shards,
-		ids:    ids,
-		fn:     fn,
+		shards:   shards,
+		ids:      ids,
+		shardMap: shardMap,
+		fn:       fn,
 	}
 }
 
 func (s *shardSet) Lookup(identifier ts.ID) uint32 {
 	return s.fn(identifier)
 }
 
+func (s *shardSet) LookupStateByID(shardID uint32) (shard.State, error) {
```

Comment:
Might be better to have the method to return the Shard instead of just the State? in case there are use cases in the future that also want to access the SourceID field